slyd/app/components/web-document-js/component.js

```diff
@@ -251,14 +255,56 @@ export default WebDocument.extend({
         if(!evt.target || !evt.target.nodeid) {
             return;
         }
+        var interaction =  interactionEvent(evt);
         this.get('ws').send({
             _meta: {
                 spider: this.get('slyd.spider'),
                 project: this.get('slyd.project'),
             },
             _command: 'interact',
-            interaction: interactionEvent(evt)
+            interaction: interaction
+        });
+        this.saveAction(interaction, evt);
+    },
+
+    saveAction: function (interactionEvent, nativeEvent) {
+        if(!experiments.enabled('page_actions')) {
+            return;
+        }
+        var pageActions = this.get('pageActions');
+        var type = interactionEvent.type;
+
+        // Filter actions we are not interested in
+        if (!pageActions || (type !== 'click' && type !== 'input' && type !== 'change')) {
+            return null; // We don't record that kind of actions
+        }
+        var target = nativeEvent.target;
+        if ((type === 'click' && $(target).is('option,select,input:text,body,textarea,html')) || // Ignore click events in some elements
+            (type === 'change' && !$(target).is('select'))){ // We only care about change events in select elements
+            return null;
+        }
+
+        // If we are inputting more text into a field, or changing again a select make it only one interaction
+        if((type === 'input' || type === 'change') && pageActions.length){
+            var lastAction = pageActions[pageActions.length-1];
+            if(lastAction.type === 'set' && lastAction.target === target && !lastAction._edited){
+                lastAction.set('value', $(target).val());
+                return;
+            }
+        }
+
+        // Record the action
+        var nonmatches = this.getIframe().find('*').not(target);
```

Comment:
It was fast in practice.

Unfortunately it uses jQuery indexOf function for that, which has linear complexity, so it's probably O(M*N) where M is the number of selectors that needs to be tested (100?) and M the number of element nodes in the page (<100000), so can't grow too big (10^8).

Of course if could be improved a lot by modifying selector gadget.